File: example/ios/Runner/NativeControlsViewFactory.swift
import Flutter
import UIKit

private let channelName = "marionette_example/native_controls"

class NativeControlsViewFactory: NSObject, FlutterPlatformViewFactory {
  private var messenger: FlutterBinaryMessenger

  init(messenger: FlutterBinaryMessenger) {
    self.messenger = messenger
    super.init()
  }

  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    return NativeControlsPlatformView(frame: frame, messenger: messenger)
  }

  func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    return FlutterStandardMessageCodec.sharedInstance()
  }
}

class NativeControlsPlatformView: NSObject, FlutterPlatformView {
  private let stackView: UIStackView
  private let channel: FlutterMethodChannel

  init(frame: CGRect, messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)

    stackView = UIStackView(frame: frame)
    stackView.axis = .vertical
    stackView.spacing = 12
    stackView.alignment = .fill
    stackView.distribution = .fill
    stackView.backgroundColor = UIColor(red: 0.91, green: 0.96, blue: 0.91, alpha: 1)
    stackView.layoutMargins = UIEdgeInsets(top: 16, left: 16, bottom: 16, right: 16)
    stackView.isLayoutMarginsRelativeArrangement = true

    let label = UILabel()
    label.text = "Native label"
    label.accessibilityIdentifier = "native_label"
    stackView.addArrangedSubview(label)

    let field = UITextField()
    field.placeholder = "Type here (native)"
    field.borderStyle = .roundedRect
    field.accessibilityIdentifier = "native_input"
    stackView.addArrangedSubview(field)

    let button = UIButton(type: .system)
    button.setTitle("Tap me (native)", for: .normal)
    button.accessibilityIdentifier = "native_button"
    stackView.addArrangedSubview(button)

    super.init()

    button.addTarget(self, action: #selector(nativeButtonTapped), for: .touchUpInside)
  }

  @objc private func nativeButtonTapped() {
    channel.invokeMethod("nativeButtonTapped", arguments: nil)
  }

  func view() -> UIView {
    return stackView
  }
}

File: example/android/app/src/main/kotlin/com/example/counter/NativeControlsViewFactory.kt
package com.example.counter

import android.content.Context
import android.graphics.Color
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

private const val channelName = "marionette_example/native_controls"

class NativeControlsViewFactory(
    private val messenger: BinaryMessenger,
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {
    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        return NativeControlsPlatformView(context, messenger)
    }
}

private class NativeControlsPlatformView(
    context: Context,
    messenger: BinaryMessenger,
) : PlatformView {
    private val channel = MethodChannel(messenger, channelName)

    private val root: LinearLayout =
        LinearLayout(context).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(32, 32, 32, 32)
            setBackgroundColor(Color.parseColor("#E8F5E9"))

            addView(
                TextView(context).apply {
                    text = "Native label"
                    contentDescription = "native_label"
                    textSize = 18f
                },
            )

            addView(
                EditText(context).apply {
                    hint = "Type here (native)"
                    contentDescription = "native_input"
                    setTextColor(Color.BLACK)
                },
            )

            addView(
                Button(context).apply {
                    text = "Tap me (native)"
                    contentDescription = "native_button"
                    setOnClickListener {
                        channel.invokeMethod("nativeButtonTapped", null)
                    }
                },
            )
        }

    override fun getView(): View = root

    override fun dispose() {}
}
